---
title: "picrust-KO-but/bile"
author: "Madi"
date: "2023-03-24"
output:
  html_document: default
  pdf_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(cowplot)
library(magrittr)
library(qiime2R)
library(tidyverse)
library(viridis)
```

**KO filtering and Stats functions**
```{r}
## allows you to filter your biom table for specified KOs and by the taxonomic level of interest
## can feed this output into func_abun_sum 
ko_filter <- function(biom, ko_list, taxonomy = 'family') { 
  biom %>% 
  select(sample, ko, taxon_function_abun, study, diet, day_post_inf, 
         any_of(taxonomy)) %>% 
    filter(ko %in% ko_list) -> filtered_biom
  return(filtered_biom)
}


my_dunn_test <- function(long_ko_table) {
  long_ko_table %>% 
    group_by(ko) %>% 
    dunn_test(preds~diet, p.adjust.method = 'bonferroni') %>% 
    add_xy_position(scales = 'free_y') -> dunn_output
  return(dunn_output)
}

## overall taxon grouping and sum of that taxon's abundance for a particular KO 
func_abun_sum <- function(filtered_biom, taxonomy='family') {
  filtered_biom %>% 
   group_by(sample, ko, study, diet, day_post_inf, .data[[taxonomy]] ) %>% 
   summarize(taxon_function_abun = sum(taxon_function_abun)) %>% 
   ungroup() -> filtered_biom_sum
  return(filtered_biom_sum)
}
```


**Unstratified KO metagenome predictions and contributions**
```{r}
## this is already in long format, that makes me happy :) 
kegg_fp <- '~/elena/picrust2_out_pipeline2/KO_metagenome_out/but_bile_contrib.tsv'
kometa_contrib <- read_tsv(file = kegg_fp)
names(kometa_contrib)[names(kometa_contrib) == 'function'] <- 'ko'
```

**Reading in taxonomy from original qiime analysis**
```{r}
taxonomy_elena <- read_qza('taxonomy_slva_mergedAim1.qza')$data %>% 
  parse_taxonomy() %>% 
  as_tibble(rownames = 'taxon') %>% 
  rename_all(tolower)

## joining picrust outputs with the qiime taxonomic analysis by asv (also called taxon here)
kometa_contrib %>% 
  left_join(taxonomy_elena, by = 'taxon') -> kometa_contrib

```

**Metadata file**
```{r}
telena_meta <- read_tsv('merged_metadata1.tsv')
elmcols <- colnames(telena_meta)
elena_meta <- read_tsv('merged_metadata1.tsv', skip = 2, col_names = elmcols) %>% 
  rename_all(tolower)

names(elena_meta)[names(elena_meta) == '#sampleid'] <- 'sample'

## pulling only the needed columns out of the metadata file to merge with taxonomy 
elena_meta %>% 
  select(sample, diet, study, day_post_inf) -> elena_meta_red
```

**Biom table construction**
Here I'm just merging the overall taxonomy table with the needed columns from the original metadata file so I can graph my needed variables. Variables that I want are the time points, diet, sample number, KO, taxonomy (whichever level wanted) and the taxon functional abundance. 
```{r}
kometa_contrib %>% 
  left_join(elena_meta_red, by = 'sample') -> kometa_contrib
```



**Ggplot construction for butyrate production**
Yellow dotted line on the graph indicates when the mice started receiving oral broad-spectrum antibiotics (should expect to see diversity go down) (day -3). 
Blue dotted line on the graph indicates when the mice received intravenous antibiotics and were gavaged with C. diff (day 0). 
```{r, fig.height=6, fig.width=14, warning=FALSE}
taxonomy_grouping <- 'class'
buty_kos <- c('K00929', 'K01034')

##putting the biom table through a filter for the KOs wanted (above)
## also groups the table by the taxonomic level of interest (must be specified)
elena_filtered_but <- ko_filter(biom = kometa_contrib, ko_list = buty_kos, 
                                taxonomy = taxonomy_grouping)

## this equation sums all of the taxon functional abundances for the KOs of interest so there
## aren't any repeated points in the ggplot (collapses them per bacteria for each KO spec.)
el_but_filt_sum <- func_abun_sum(elena_filtered_but, 
                              taxonomy = taxonomy_grouping)

## this is setting up the labels for my 'by KO' facet grids to say the actual enzyme names and not 
## just the KO number 
buty_labs <- c('Butyrate\nKinase', 'Butyryl-CoA\nTransferase')
names(buty_labs) <- c('K00929', 'K01034')

## my actual ggplot (also known as my child)
## put your fill/color variable in geom_point/jitter so that it doesn't mess up your geom_smooth line calculations
el_but_filt_sum %>% 
  filter(!is.na(diet)) %>% 
  ggplot(aes(x = day_post_inf, y = taxon_function_abun)) +
    geom_smooth(method = 'loess', se = FALSE, color = 'black') +
    geom_vline(xintercept = -3, linetype = 'dashed', color = 'blue', size = 0.2) +
    geom_vline(xintercept = 0, linetype = 'dashed', color = 'purple', size = 0.2) +
    geom_jitter(aes(fill = class),
                width = 0.1, height = 0, 
                alpha = 0.7, pch = 21) +
    scale_x_continuous(breaks = c(-15, -8, -3, 0, 3)) + 
    scale_fill_brewer(palette = 'Dark2') +
    facet_grid(ko~diet, scales = 'free', labeller = labeller(ko = buty_labs)) +
    theme_bw(base_size = 12) +
    theme(legend.text = element_text(size = 8.5)) +
    guides(color = guide_legend(override.aes = list(size = 0.9))) + 
    ggtitle('Butyrate Potential Over Time') +
    xlab('Days Relative to Infection') +
    ylab('KO Counts') +
    labs(fill = 'Class')
```

**Ggplot construction for secondary bile acid production**
baiH and baiI are enzymes of interest 
baiH ko = K15873
baiI ko = K15874
Want to go through and add zeroes for the diets where the secondary bile acid production disappears (extrapolation?)
```{r, fig.height=5, fig.width=9, warning=FALSE}
bile_acid_kos <- c('K15873', 'K15874')
filtered_bile_acid <- ko_filter(biom = kometa_contrib, ko_list = bile_acid_kos, 
                                taxonomy = 'genus')

bile_acid_sum <- func_abun_sum(filtered_bile_acid, taxonomy = 'genus')

bile_acid_labs <- c('baiH', 'baiI')
names(bile_acid_labs) <- c('K15873', 'K15874')
  
bile_acid_sum %>% 
  filter(!is.na(diet)) %>% 
  ggplot(aes(x = day_post_inf, y = taxon_function_abun, fill = genus)) +
    geom_smooth(method = 'loess', se = FALSE, color = 'black', fullrange = TRUE) +
    geom_vline(xintercept = -3, linetype = 'dotted', color = 'purple', size = 0.5) +
    geom_vline(xintercept = 0, linetype = 'dashed', color = 'red', size = 0.5) +
    geom_jitter(width = 0.1, height = 0, alpha = 0.4, pch=21) +
    scale_fill_viridis(option = 'D', discrete = TRUE) +
    facet_grid(ko~diet, scales = 'free_y', labeller = labeller(ko = bile_acid_labs)) +
    theme_bw(base_size = 12) +
    theme(legend.text = element_text(size = 9.5), 
          legend.position = 'bottom') +
    guides(fill = guide_legend(override.aes = list(size = 2.5))) + 
    ggtitle('Secondary Bile Acid Potential Over Time') +
    xlab('Days Relative to Infection') +
    ylab('KO Counts') +
    labs(fill = 'Genus')
```


**Rearranging Bile Acid Table**
I needed to move my bile acid table around to include the zero values so they would be seen on the ggplot I constructed (because it's weird to see all the points just suddenly drop off like in the graph above). So, we converted the table from long format to wide format (to show all the NAs that weren't included before in the long format), filled all the NAs with zeroes and then put the table back into long format to use for my ggplot. 
We had to tell the filter to exclude any NAs for the days post infection (there were a couple). 
We also needed to tell tidyverse that day_post_inf was numeric instead of categorical so that my geom_smooth would work (i.e. there would be an actual line on the graph). 
```{r}
bile_acid_sum %>% 
  filter(!is.na(day_post_inf)) %>% 
  spread(day_post_inf, taxon_function_abun, fill = 0) %>% 
  gather(-sample, -ko, -study, -diet, -genus, 
         key = day_post_inf, value = taxon_function_abun) %>% 
  mutate(day_post_inf = as.numeric(day_post_inf)) -> bile_acid_sum_w_0
```

**GGplot for Bile Acid Synthesis (part two) - corrected**
I needed to create a new ggplot where my geom_smooth lines would extend for all the zero values instead of stopping halfway through the graph. I used the table I created above to generate a new ggplot that had the desired modifications. 
```{r, fig.height=5, fig.width=9, warning=FALSE}
bile_acid_sum_w_0 %>% 
  filter(!is.na(diet)) %>% 
  ggplot(aes(x = day_post_inf, y = taxon_function_abun)) +
    geom_smooth(se = FALSE, color = 'black') +
    geom_vline(xintercept = -3, linetype = 'dashed', color = 'blue', size = 0.2) +
    geom_vline(xintercept = 0, linetype = 'dashed', color = 'purple', size = 0.2) +
    geom_jitter(width = 0.1, height = 0, alpha = 0.7, pch=21,
                aes(fill = genus)) +
    scale_x_continuous(breaks = c(-15, -8, -3, 0, 3)) + 
    scale_fill_brewer(palette = 'Dark2') +
    facet_grid(ko~diet, scales = 'free_y', labeller = labeller(ko = bile_acid_labs)) +
    theme_bw(base_size = 12) +
    theme(legend.text = element_text(size = 9.5), 
          legend.position = 'bottom') +
    guides(fill = guide_legend(override.aes = list(size = 2.5))) + 
    ggtitle('Secondary Bile Acid Potential Over Time') +
    xlab('Days Relative to Infection') +
    ylab('KO Counts') +
    labs(fill = 'Genus')
```

